### ISSUE TITLE
Rule proposal check void-covariant return types of implemented interfaces and extended classes

### ISSUE THREAD

**Issue by rendaw** [typescript-eslint/typescript-eslint#1744]
**What code were you trying to parse?**

```ts
interface X {
  a(): void;
}

class Y implements X {
  async a(): Promise<void> {
  }
}
```

**What did you expect to happen?**

This should (configurably) be an error, especially concerning floating promises.

**What actually happened?**

No error.

**Versions**

```
    "@typescript-eslint/eslint-plugin": "^2.24.0",
    "@typescript-eslint/eslint-plugin-tslint": "^2.24.0",
    "@typescript-eslint/parser": "^2.24.0",
    "eslint": "^6.8.0",
    "eslint-config-prettier": "^6.10.0",
    "typescript": "^3.8.3"
```

**Other**

This might be related: https://github.com/microsoft/TypeScript/issues/4544

**Comment by bradzacher** [comment]
What are you asking for here? Could you clarify?
Is there a bug with a rule, or are you asking for an enhancement to a rule?

**Comment by rendaw** [comment]
I'd like typescript-eslint to check for this condition and warn/error about it.

I'm not sure I can say whether it's a bug dealing with a corner case or enhancement.  IMO the floating-promises rule should catch this (promise return where base class method is void return), but a separate rule to check for non-promise return types may also be useful (ex: cases where the returned value represents an opened resource that could leak).

**Comment by rendaw** [comment]
Whoops, noticed I had the example wrong (missing `implements`).

**Comment by bradzacher** [comment]
`floating-promises` is about calling a function which produces a promise, without any `await` or `then`. So there is a part of this that might make sense in that rule; passing a promise returning function into a position that's expecting a `void` return.
```ts
function foo(fn: () => void) {}
foo(() => Promise.resolve());
```

----

In terms of checking for implements assignability, that would be its own rule, as there are various problems with the variance of interface methods. Though those should probs be fixed up in typescript itself.

**Comment by ark120202** [comment]
It sounds quite close to [`no-misused-promises`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-misused-promises.md)'s `checksVoidReturn` part

**Comment by bradzacher** [comment]
oh yeah it does, thanks for reminding me.
it's exactly what `no-misused-promises` `checksVoidReturn` does.

So that leaves just the last part - of checking variance of methods from implements/extends

**Comment by rendaw** [comment]
I'm no expert, but that analysis sounds sounds correct to me.

Maybe I'm stupid, but just got bit by this again.  In any case, thanks for taking a look!

**Comment by ypresto** [comment]
Today I found possible false-negative for `no-misused-promises`

```ts
// onClick is `type EventHandler<E extends SyntheticEvent<any>> = { bivarianceHack(event: E): void }["bivarianceHack"];`

// Not warned
const link = <a onClick={e => asyncFunc()} />

interface Foo {
  onClick: () => void
}

const foo: Foo = {
  // Not warned
  onClick: async () => null,
}
```

Are these covered by this issue?

**Comment by JoshuaKGoldberg** [comment]
Ah, #8765 actually add an _implemented interfaces and extended classes_ check to [`@typescript-eslint/no-misused-promises`](https://typescript-eslint.io/rules/no-misused-promises).

> [https://github.com/typescript-eslint/typescript-eslint/issues/1744#issuecomment-636412093] Are these covered by this issue?

The `onClick: async () => null,` case is now! But not the `onClick={e => asyncFunc()}` case. That's in #2988.

Closing this issue out as fixed by #8765. #2988 remains open for passing callbacks. 👍
